internal/worker/apiremotecaller/worker.go

```diff
@@ -0,0 +1,347 @@
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Ask for the current server details now that we're subscribed.
+	detailsRequest := apiserver.DetailsRequest{
+		Requester: "api-remote-worker",
+		LocalOnly: true,
+	}
+	if _, err := cfg.Hub.Publish(apiserver.DetailsRequestTopic, detailsRequest); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	err = catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{
+			w.runner,
+		},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+// GetAPIRemotes returns the current API connections. It is expected that
+// the caller will call this method just before making an API call to ensure
+// that the connection is still valid. The caller must not cache the connections
+// as they may change over time.
+func (w *remoteWorker) GetAPIRemotes() []RemoteConnection {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	return w.apiRemotes
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *remoteWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *remoteWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *remoteWorker) loop() error {
+	defer w.unsubServerDetails()
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+
+		case change := <-w.changes:
+			w.cfg.Logger.Debugf("remoteWorker API server change: %v", change)
+
+			// Locate the current workers, so we can remove any workers that are
+			// no longer required.
+			current := w.runner.WorkerNames()
+
+			required := make(map[names.Tag]RemoteConnection)
+			for target, addresses := range change.servers {
+
+				server, err := w.newRemoteServer(target, addresses)
+				if err != nil {
+					w.cfg.Logger.Errorf("failed to start remote worker for %q: %v", target, err)
+					return errors.Trace(err)
+				}
+
+				// Always update the server with the latest addresses, even if
+				// it was just created. This is to ensure that the server is
+				// always up to date with the latest addresses.
+				server.UpdateAddresses(addresses)
+
+				required[target] = server
+			}
+
+			// Walk over the current workers and remove any that are no longer
+			// required.
+			for _, s := range current {
+				target, err := names.ParseTag(s)
+				if err != nil {
+					// This really should never happen, but if it does, bounce
+					// the worker and start again.
+					w.cfg.Logger.Errorf("failed to parse tag %q: %v", s, err)
+					return errors.Trace(err)
+				}
+
+				if _, ok := required[target]; ok {
+					continue
+				}
+
+				w.cfg.Logger.Debugf("remote worker %q no longer required", target)
+				w.stopRemoteServer(ctx, target)
+			}
+
+			w.cfg.Logger.Debugf("remote workers updated: %v", required)
+
+			remotes := make([]RemoteConnection, 0, len(required))
+			for _, remote := range required {
+				remotes = append(remotes, remote)
+			}
```

Comment:
Go was once a great language, now there are two ways to do this.. and this new way sucks (because it doesn't preallocate the slice backing array).

```suggestion
			remotes := slices.Collect(maps.Values(required))
```